README.md
# StepwiseLH
Package to generate life history probability distributions from an estimate of maximum length.

## GitHub Disclaimer
This repository is a scientific product and is not official communication of the National Oceanic and Atmospheric Administration, or the United States Department of Commerce. All NOAA GitHub project code is provided on an ‘as is’ basis and the user assumes responsibility for its use. Any claims against the Department of Commerce or Department of Commerce bureaus stemming from the use of this GitHub project will be governed by all applicable Federal law. Any reference to specific commercial products, processes, or services by servicemark, trademark, manufacturer, or otherwise, does not constitute or imply their endorsement, recommendation or favoring by the Department of Commerce. The Department of Commerce seal and logo, or the seal and logo of a DOC bureau, shall not be used in any manner to imply endorsement of any commercial product or activity by DOC or the United States Government.

# Package Installation

```
library(devtools)
install_github("PIFSCstockassessments/StepwiseLH")
```

# Usage

`StepwiseLH`'s primary method is `Get_distributions`. *TODO: [Short description]*. `Lmax` is the 99th percentile of length in a typical survey or catch dataset.


The following example generates random distributions of _Acanthuridae_ (Surgeonfishes) governed by Life History parameters, `Lmax` mean and standard deviations, as well as surviorship at max age.
```
library(StepwiseLH)
Data <- Get_distributions(Family="Acanthuridae", Lmax.mean=600, Lmax.SD=50, M_method=0.04)
```
`Data` is the **data.table** object returned from `Get_distributions`:
```
      Lmax            Linf             K                 A0            Lmat            Amat              M                Amax       
 Min.   :401.2   Min.   :356.1   Min.   :0.05388   Min.   :-0.6   Min.   :250.4   Min.   : 0.409   Min.   :0.01252   Min.   : 16.55  
 1st Qu.:566.1   1st Qu.:532.9   1st Qu.:0.19818   1st Qu.:-0.6   1st Qu.:412.8   1st Qu.: 3.756   1st Qu.:0.08833   1st Qu.: 26.41  
 Median :599.1   Median :570.4   Median :0.26766   Median :-0.6   Median :456.1   Median : 5.568   Median :0.10490   Median : 30.68  
 Mean   :599.3   Mean   :571.0   Mean   :0.29653   Mean   :-0.6   Mean   :460.6   Mean   : 6.575   Mean   :0.10479   Mean   : 33.11  
 3rd Qu.:633.6   3rd Qu.:608.0   3rd Qu.:0.36577   3rd Qu.:-0.6   3rd Qu.:502.6   3rd Qu.: 8.218   3rd Qu.:0.12188   3rd Qu.: 36.44  
 Max.   :795.8   Max.   :776.2   Max.   :1.32707   Max.   :-0.6   Max.   :791.4   Max.   :53.919   Max.   :0.19445   Max.   :257.05  
                                                                                  NA's   :117                                        
          Lmax     Linf          K   A0     Lmat      Amat          M     Amax
   1: 614.3950 570.0183 0.25820404 -0.6 396.4040  4.004236 0.13282306 24.23432
   2: 512.2795 484.6708 0.26622224 -0.6 367.7741  4.742076 0.13572667 23.71587
   3: 682.3660 673.2720 0.26642125 -0.6 637.8082 10.448808 0.06272129 51.32031
   4: 626.8624 611.9785 0.21054987 -0.6 497.2772  7.352349 0.13695462 23.50323
   5: 401.2273 356.0943 0.35028486 -0.6 295.5570  4.458554 0.07497098 42.93496
  ---                                                                         
4996: 608.2304 597.3491 0.52836143 -0.6 458.5285  2.161971 0.11358994 28.33768
4997: 634.4598 590.6622 0.09441384 -0.6 440.5053 13.905965 0.12567246 25.61322
4998: 625.6482 593.5151 0.17901557 -0.6 554.4157 14.593953 0.08575178 37.53713
4999: 545.6864 483.2731 0.20089035 -0.6 375.8214  6.884384 0.09015481 35.70387
5000: 637.0055 646.0454 0.22407480 -0.6 559.9136  8.392491 0.09547127 33.71565
```

### Available families

The following `Family` values available for `Get_distributions`:


| Family             | Common Name                      |
| ------------------ | -------------------------------- |
| `Acanthuridae`     | Surgeonfishes                    |
| `Carangidae`       | Jacks                            |
| `Lethrinidae`      | Emperors                         |
| `Lutjanidae`       | Snappers                         |
| `Mullidae`         | Goatfishes                       |
| `Scaridae`         | Parrotfishes                     |

## References 
TODO




NEWS.md

# StepwiseLH ver 1.1

## Major Changes

* Added `GenerateRandom` to `NAMESPACE`


# StepwiseLH ver 1.0

Initial Version


R/Get_randomLH.R
Get_randomLH <- function(aFunction, aDist, coef, cov, error, x_value1, x_value2){

  aRandomCoef <- mvrnorm(n=1, mu=coef,Sigma=cov)

  # Expected value
  if(aFunction=="Linear")     { E <- aRandomCoef[1]+aRandomCoef[2]*x_value1       }
  if(aFunction=="Expo")       { E <- aRandomCoef[1]+aRandomCoef[2]*x_value1       }
  if(aFunction=="Power")      { E <- aRandomCoef[1]+aRandomCoef[2]*log(x_value1)  }
  if(aFunction=="Power_2var") { E <- aRandomCoef[1]+aRandomCoef[2]*log(x_value1)+aRandomCoef[3]*log(x_value2) }

  # Error distribution
  if(aDist=="Normal")  {  aRandom <- rtruncnorm(n=1,a=0,b=Inf,mean=E,sd=error)    }

  if(aDist=="Lognormal")
  {
    if(aFunction=="Linear")                                               { aRandom <- rlnorm(n=1,log(E),error) }
    else if(aFunction=="Expo"|aFunction=="Power"|aFunction=="Power_2var") { aRandom <- rlnorm(n=1,E,error)      }
  }

  return (aRandom);
}


R/GenerateRandom.R
#' Generate Random Distrbution
#'
#' Generates random deviates. \code{aDist} determinines normal or lognormal
#' distribution of the random generation function.
#'
#' @param n_iter Number of oberservations
#' @param aDist Type of Distribution: Normal or Lognormal.
#' @param par1 Vector of Means
#' @param par2 Vector of Standard Deviation
#'
#' @seealso \code{\link[truncnorm]{rtruncnorm}}
#' @seealso \code{\link[stats]{rlnorm}}
#'
#' @importFrom truncnorm rtruncnorm
#' @importFrom stats rlnorm
#' @export
GenerateRandom <- function(n_iter, aDist, par1, par2){

  #require(truncnorm)
  if(aDist=="Normal")    {  aRandom <- rtruncnorm(n=n_iter,a=0,b=Inf,mean=par1,sd=par2)  }
  if(aDist=="Lognormal") {  aRandom <- rlnorm(n=n_iter,par1,par2)              }

  return (aRandom);
}


R/Get_coefficients.R
#' Get coefficients
#'
#' Retrieves Life History values for Family
#'
#' @param Family Taxaon Name
#' @param Lmax.mean Mean of 'largest locally measured length'.
#'
#' @export
Get_coefficients <- function(Family, Lmax.mean){

  if(Family=="Acanthuridae")
  {
    LinfFunc   <- "Linear"
    LinfDist   <- "Normal"
    Linf_coef  <- rbind(-39.421486,1.018236)
    Linf_cov   <- rbind(c(136.5050105, -0.345017509),
                        c( -0.3450175,  0.001037638))
    Linf_error <- 18.073072

    KFunc    <- "Expo"
    KDist    <- "Lognormal"
    K_coef   <- rbind(-0.27823534,-0.00179617)
    K_cov    <- rbind(c( 3.628268e-02,-8.492274e-05),
                      c(-8.492274e-05, 2.423078e-07))
    K_error  <- 0.41955247

    A0Dist   <- "Fixed"
    A0_coef  <- rbind(-0.6, -999, -999)
    A0_cov   <- rbind(-999)

    MFunc    <- "Linear"
    MDist    <- "Normal"
    M_coef   <- rbind(0.09768,0);
    M_cov    <- rbind(c(1.994e-5,0),c(0,0))
    M_error  <- 0.02277

    MatFunc  <- "Linear"
    MatDist  <- "Lognormal"
    Mat_coef <- rbind(-12.551248,0.827516)
    Mat_cov  <- rbind(c( 1.944283e+02,-6.928236e-01),
                      c(-6.928236e-01, 3.050065e-03))
    Mat_error<- 0.102
  }
  else if(Family=="Carangidae")
  {

    if(Lmax.mean<=900){
      LinfFunc  <- "Linear"
      LinfDist  <- "Lognormal"
      Linf_coef <- rbind(27.997851,1.003558)
      Linf_cov  <- rbind(c( 1.650201e+03,-3.3814631668),
                         c(-3.3814631668, 0.0081213854))
      Linf_error<- 0.1302
    }else{
      LinfFunc  <- "Linear"
      LinfDist  <- "Lognormal"
      Linf_coef <- rbind(-27.938,1.196)
      Linf_cov  <- rbind(c(8.624730e-09,9.937220e-06),
                         c(9.937220e-06,1.145423e-02))
      Linf_error<- 0.24
    }

    KFunc    <- "Expo"
    KDist    <- "Lognormal"
    K_coef   <- rbind(-1.10386285,-0.00045998)
    K_cov    <- rbind(c( 1.772833e-02,-1.721886e-05),
                      c(-1.721886e-05, 2.206721e-08))
    K_error  <- 0.35885624

    A0Dist   <- "Fixed"
    A0_coef  <- rbind(-0.6,-999,-999)

    MFunc    <- "Power_2var"
    MDist    <- "Lognormal"
    M_coef   <- rbind(2.267174,0.030239,-0.515081)
    M_cov    <- rbind(c( 7.638689e-01,-2.137078e-03,-1.179220e-01),
                      c(-2.137078e-03, 2.197155e-02, 5.211300e-03),
                      c(-1.179220e-01, 5.211300e-03, 1.939093e-02))
    M_error  <- 0.307

    MatFunc  <- "Linear"
    MatDist  <- "Normal"
    Mat_coef <- rbind(33.576092,0.572904)
    Mat_cov  <- rbind(c(542.3375676,-0.611514575),
                      c( -0.6115146, 0.000859941))
    Mat_error<- 43.96
  }
  else if(Family=="Lethrinidae")
  {

    LinfFunc  <- "Linear"
    LinfDist  <- "Normal"
    Linf_coef <- rbind(-1.441152,0.872423)
    Linf_cov  <- rbind(c(161.8062667,-0.333560243),
                       c( -0.3335602, 0.000752506))
    Linf_error<- 17.1

    KFunc    <- "Power"
    KDist    <- "Lognormal"
    K_coef   <- rbind(4.658581,-0.930113)
    K_cov    <- rbind(c( 3.124779    ,-5.257665e-01),
                      c(-5.257665e-01, 8.871238e-02))
    K_error  <- 0.438728

    A0Dist   <- "Fixed"
    A0_coef  <- rbind(-0.6,-999,-999)

    MFunc    <- "Power_2var"
    MDist    <- "Lognormal"
    M_coef   <- rbind(0.54085,0.53750,-0.28432)
    M_cov    <- rbind(c( 2.115091    , -9.768916e-02,-3.619241e-01),
                      c(-9.768916e-02,  2.494511e-02, 1.944466e-02),
                      c(-3.619241e-01,  1.944466e-02, 6.240972e-02))
    M_error  <- 0.297

    MatFunc  <- "Linear"
    MatDist  <- "Normal"
    Mat_coef <- rbind(36.013807,0.671787)
    Mat_cov  <- rbind(c(395.5265340,-1.0156028933),
                      c( -1.0156029, 0.0028603607))
    Mat_error<- 21.3
  }
  else if(Family=="Lutjanidae")
  {
    if(Lmax.mean<=500){
      LinfFunc  <- "Linear"
      LinfDist  <- "Normal"
      Linf_coef <- rbind(-5.077514,0.912436)
      Linf_cov  <- rbind(c(952.051969,-2.651956402),
                         c( -2.651956, 0.007722089))
      Linf_error<- 21.3
    }else{
      LinfFunc  <- "Linear"
      LinfDist  <- "Normal"
      Linf_coef <-rbind(-81.77184,1.05954)
      Linf_cov  <-rbind(c(6711.302438,-8.30116144),
                        c(  -8.301161, 0.01076043))
      Linf_error<- 85.86606
    }

    KFunc    <- "Power"
    KDist    <- "Lognormal"
    K_coef   <- rbind(5.225192,-1.056897)
    K_cov    <- rbind(c( 7.595088e-01,-1.196958e-01),
                      c(-1.196958e-01, 1.896458e-02))
    K_error  <- 0.397114

    A0Dist   <- "Fixed"
    A0_coef  <-rbind(-0.6,-999,-999)

    MFunc    <- "Power_2var"
    MDist    <- "Lognormal"
    M_coef   <- rbind(2.0379245,-0.0086432,-0.6306451)
    M_cov    <- rbind(c( 1.723808    ,-1.386993e-01,-3.004497e-01),
                      c(-1.386993e-01, 2.507150e-02, 2.729170e-02),
                      c(-3.004497e-01, 2.729170e-02, 5.317334e-02))
    M_error  <- 0.3929

    MatFunc  <- "Linear"
    MatDist  <- "Lognormal"
    Mat_coef <- rbind(63.875705,0.519023)
    Mat_cov  <- rbind(c(423.325388756,-8.530914e-01),
                      c( -0.853091438, 2.161344e-03))
    Mat_error<- 0.148013
  }
  else if(Family=="Mullidae")
  {
    LinfFunc  <- "Linear"
    LinfDist  <- "Normal"
    Linf_coef <- rbind(-9.340027,1.158293)
    Linf_cov  <- rbind(c(518.672234,-1.967218231),
                       c( -1.967218, 0.007936577))
    Linf_error<- 16.7

    KFunc    <- "Linear"
    KDist    <- "Normal"
    K_coef   <- rbind(0.466636,0)
    K_cov    <- rbind(c(0.0027,0),c(0,0))
    K_error  <- 0.1727

    A0Dist   <- "Fixed"
    A0_coef  <- rbind(-0.6,-999,-999)

    MFunc    <- "Linear"
    MDist    <- "Lognormal"
    M_coef   <- rbind(0.6744,0)
    M_cov    <- rbind(c(0.00689,0),c(0,0))
    M_error  <- 0.348

    MatFunc  <- "Linear"
    MatDist  <- "Normal"
    Mat_coef <- rbind(49.32657,0.414815)
    Mat_cov  <- rbind(c(344.831162   ,-1.135673725),
                      c( -1.135673725, 0.004357041))
    Mat_error<- 22.09
  }
  else if(Family=="Scaridae")
  {
    LinfFunc  <- "Linear"
    LinfDist  <- "Normal"
    Linf_coef <- rbind(0.190650,0.849181)
    Linf_cov  <- rbind(c(378.6815365,-0.848459796),
                       c( -0.8484598, 0.002086915))
    Linf_error<- 23.2

    KFunc    <- "Expo"
    KDist    <- "Lognormal"
    K_coef   <- rbind(0.68111900,-0.00353955)
    K_cov    <- rbind(c( 2.681769e-02,-6.586431e-05),
                      c(-6.586431e-05, 1.933287e-07))
    K_error  <- 0.33085776

    A0Dist   <- "Fixed"
    A0_coef  <- rbind(-0.6,-999,-999)

    MFunc    <- "Power_2var"
    MDist    <- "Lognormal"
    M_coef   <- rbind(-0.768542,0.646086,-0.039457)
    M_cov    <- rbind(c( 3.465647    ,-2.579116e-01,-6.015445e-01),
                      c(-2.579116e-01, 3.699427e-02, 4.640341e-02),
                      c(-6.015445e-01, 4.640341e-02, 1.046714e-01))
    M_error  <- 0.241926

    MatFunc  <- "Linear"
    MatDist  <- "Normal"
    Mat_coef <- rbind(-7.967278,0.713637)
    Mat_cov  <- rbind(c(183.0398307,-0.42502341),
                      c( -0.4250234, 0.00114699))
    Mat_error<- 22
  }
  else{
    stop("Invalid Family value")
  }


  return( list(LinfFunc=LinfFunc,LinfDist=LinfDist,Linf_coef=Linf_coef,Linf_cov=Linf_cov,Linf_error=Linf_error,
               KFunc=KFunc,KDist=KDist,K_coef=K_coef,K_cov=K_cov,K_error=K_error,
               A0Dist=A0Dist,A0_coef=A0_coef,
               MFunc=MFunc,MDist=MDist,M_coef=M_coef,M_cov=M_cov,M_error=M_error,
               MatFunc=MatFunc,MatDist=MatDist,Mat_coef=Mat_coef,Mat_cov=Mat_cov,Mat_error=Mat_error
  )  )


} # End of function


R/Get_distributions.R
#' Family taxon distributions
#'
#' Returns Life History distributions based on a selected Family taxon.
#'
#' @param Family_Input Selection of available family taxon.
#' @param Lmax.mean Mean of 'largest locally measured length'.
#' @param Lmax.SD Standard Deviation of 'largest locally measured length'.
#' @param M_method Select how M is calculated. Default is 0.04.
#' @param n_iter Number of obervations. Default is 5000.
#' @param dist_summary Option for descriptive statistics and graphs. Default is TRUE.
#'
#' @import data.table
#' @import truncnorm
#' @import MASS
#' @importFrom graphics hist par
#' @export

Get_distributions <- function(Family_Input, Lmax.mean, Lmax.SD, M_method=0.04, n_iter=5000, dist_summary=TRUE){

  if(is.numeric(Family_Input)){
  Family_Names <- c("Acanthuridae","Carangidae","Lethrinidae","Lutjanidae","Mullidae","Scaridae")
  Family       <- Family_Names[Family_Input]
  } else {Family <- Family_Input}

  coefs <- Get_coefficients(Family, Lmax.mean)

  Lmax <- vector(length=n_iter); Linf <- vector(length=n_iter); K    <- vector(length=n_iter); A0<- vector(length=n_iter)
  M    <- vector(length=n_iter); Amax <- vector(length=n_iter); Lmat <- vector(length=n_iter); Amat <- vector(length=n_iter)

  for(i in 1:n_iter)
  {
    Lmax[i] <- rtruncnorm(1,a=0,b=Inf,Lmax.mean,Lmax.SD)
    Linf[i] <- Get_randomLH(coefs$LinfFunc,coefs$LinfDist,coefs$Linf_coef,coefs$Linf_cov,coefs$Linf_error, Lmax[i])
    K[i]    <- Get_randomLH(coefs$KFunc,coefs$KDist,coefs$K_coef,coefs$K_cov,coefs$K_error, Linf[i])
    A0[i]   <- coefs$A0_coef[1]
    M[i]    <- Get_randomLH(coefs$MFunc,coefs$MDist,coefs$M_coef,coefs$M_cov,coefs$M_error, K[i], Lmax[i])
    Amax[i] <- -log(0.05)/M[i]

    # Select how M is calculated
    if(is.numeric(M_method)) {
      M[i] <- -log(M_method)/Amax[i]        # rule-of-thumb approach where M=-log(Survivorship)/Amax see Hewitt & Hoenig (2005) equation (2). Default is S=0.04 based on Nadon et al. (2015).
    }else if(M_method=="Then_2014") {
      M[i] <- 4.899*Amax[i]^-0.916          # new equation from Then et al. (2014)
    }else if(M_method=="Pauly_Then_2014") {
      M[i] <- 4.118*K[i]^0.73*Linf[i]^-0.33 # updated Pauly (1980) equation from Then et al. (2014)
    }else{
      stop("Invalid M method")
    }

    LLambda <- Linf[i]*(1-exp(-K[i]*(Amax[i]-A0[i])))
    repeat{
    Lmat[i] <- Get_randomLH(coefs$MatFunc,coefs$MatDist,coefs$Mat_coef,coefs$Mat_cov,coefs$Mat_error, LLambda)
    if(Lmat[i]<Linf[i]) break
    }
    Amat[i] <- A0[i]-1/K[i]*log(1-Lmat[i]/Linf[i])

  }

  Out <- data.table( cbind(Lmax,Linf,K,A0,Lmat,Amat,M,Amax) )

  if(dist_summary){
    # Some descriptive statistics and graphs
    Linf <- Out[,Linf]
    K <- Out[,K]
    M <- Out[,M]
    Lmat <- Out[,Lmat]

    par(mfrow=c(2,2))
    hist(Linf)
    hist(K)
    hist(M)
    hist(Lmat)

    out_summary <- summary(Out)
    print(out_summary)
  }

  return(Out)

}
